src/main/kotlin/cacophony/grammars/Analysis.kt

```diff
@@ -4,10 +4,54 @@ import cacophony.automata.DFA
 
 typealias DFAStateReference<StateType, AtomType, ResultType> = Pair<StateType, DFA<StateType, AtomType, ResultType>>
 
-fun <StateType, SymbolType> findNullable(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-): Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>> {
-    TODO("Not implemented")
+private fun <StateType, AtomType, ResultType> reverse(
+    dfa: DFA<StateType, AtomType, ResultType>,
+): Map<StateType, Map<AtomType, Set<StateType>>> =
+    dfa
+        .getProductions()
+        .asSequence()
+        .map { (it.value to it.key.second) to it.key.first }
+        .groupBy({ it.first }, { it.second })
+        .map { it.key.first to (it.key.second to it.value.toSet()) }
+        .groupBy({ it.first }, { it.second })
+        .map { it.key to it.value.toMap() }
+        .toMap()
```

Comment:
you can extract it as utility function to `minimalization.DFAUtils.kt`, where we put all our dfa utils for now